admin/services/data-access/pupil-census-import.data.service.js

```diff
@@ -0,0 +1,72 @@
+'use strict'
+
+const { TYPES } = require('tedious')
+const moment = require('moment')
+
+const config = require('../../config')
+const pupilCensusImportDataService = {}
+const sqlService = require('./sql.service')
+const sqlPoolService = require('./sql.pool.service')
+
+/**
+ * Execute pupil data bulk import
+ * @param {Array} pupilData
+ * @param {Array} schools
+ * @param {Number} jobId
+ * @return {Promise<*>}
+ */
+pupilCensusImportDataService.sqlBulkImport = async(pupilData, schools, jobId) => {
+  const result = {}
+  const con = await sqlPoolService.getConnection()
+
+  try {
+    result.output = await bulkLoadData(con, pupilData, schools, jobId)
+  } catch (error) {
+    result.errorOutput = error
+  }
+  con.release()
+  return result
+}
+
+const bulkLoadData = (connection, pupilData, schools, jobId) => {
+  return new Promise((resolve, reject) => {
+    const bulkLoad = connection.newBulkLoad(`${config.Sql.Database}.${sqlService.adminSchema}.[pupil]`, function (error, rowCount) {
+      if (error) {
+        return reject(error)
```

Comment:
If we load 100 records in the file, and 1 record fails validation at the DB level, do the whole 100 records get rejected?  Or do 99 get inserted and 1 rejected?

Ideally, I would want to load 99 records that are good and log everything about the record that failed.  Does the service manager get enough information here to proceed even in the event of failures and incompatibilities?

I would expect to see a process that chunks up the input, which may be very large, into suitable chunks for batch processing as it simply cannot ingest an unlimited amount of data.

Reply:
@jon-shipley In case of a validation error during a bulk insertion attempt all the records get rejected. I am not sure if this is something we can control as this is driven by tedious as far as I see it.

A process that chunks up the input wouldn't be required at this stage as we are dealing with a small dataset of maximum 6000 records.

Reply:
What message does the user get then?